Add snapcast latency attribute and service

BarrettLowe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This allows users to set latency of snapcast devices. This is useful when connecting higher latency devices to the snapcast system. Users can now write automations to adjust the latency of all connected devices if a slow device joins etc.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12904

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

BarrettLowe committed:
Implemented snapcast latency attributes

homeassistant commented:
Hello @BarrettLowe,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ea8f9f5938c28cc5ba4591abf2f0403bda086489`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;34126&#x2F;commits&#x2F;ea8f9f5938c28cc5ba4591abf2f0403bda086489) - This commit has something that looks like an email address (Barrett.Lowe@etcconnect.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @BarrettLowe, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#34126).

Once you do that we will be able to review and accept this pull request.

Thanks!

BarrettLowe commented:
Can someone tell me why I'm getting errors in the codecov/patch check? Looks like it's complaining about files that I didn't touch. Also I think I fixed my cla-error

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -84,6 +86,14 @@
+                        for e in hass.data[DATA_KEY]
+                        if e.entity_id == data[ATTR_ENTITY_ID]
+                    ]
+                    await device.async_set_latency(data.get(ATTR_LATENCY))
```
Please use our new way of registering entity service handlers:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

BarrettLowe replied:
Should I change all existing service calls also?

MartinHjelmare replied:
It's ok to add a new handler just for this service and migrate the other services later.

If you want to clean them up now and can test that, it's ok too.

BarrettLowe replied:
Okay this is my first contribution so give me a bit to figure it out. Want the commits squashed or want a second commit for code fixes/cleanup?

MartinHjelmare replied:
Please just add new commits after review has started. That makes it easier to track the changes.

We will squash when merging.

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -257,6 +267,15 @@ def state(self):
         return STATE_OFF
 
+    @property
+    def state_attributes(self):
```
Remove this.

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -257,6 +267,15 @@ def state(self):
+        state_attrs = {}
+        state_attrs.update(super().state_attributes)
+        if self.latency is not None:
+            state_attrs["latency"] = self.latency
```
Move this to `device_state_attributes`.

MartinHjelmare commented:
In `.coveragerc`, exclude all snapcast modules in the snapcast package with a star:
https://github.com/home-assistant/core/blob/4bcceb4078633be10c2f144bde450e11d41d77a4/.coveragerc#L651

BarrettLowe committed:
Code review changes and Snapcast maintenance

Updated how entity services get called - now conforms to most
current method

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -58,36 +57,23 @@
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the Snapcast platform."""
 
+    _LOGGER.debug("Reached async_setup_platform")
```
Please remove this before we merge.

MartinHjelmare commented on homeassistant/components/snapcast/__init__.py:
```diff
@@ -1,50 +1,23 @@
-    )
-
-    return True
+LATENCY_SCHEMA = SERVICE_SCHEMA.extend({vol.Required(ATTR_LATENCY): cv.positive_int})
```
We can remove all the schemas, right?

MartinHjelmare replied:
The remaining constants can be put in a const.py module.

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -287,10 +282,11 @@ async def async_set_volume_level(self, volume):
 
     async def async_join(self, master):
         """Join the group of the master player."""
+        masters = [e for e in self.hass.data[DATA_KEY] if e.entity_id == master]
```
```suggestion
        master_entity = next(
            (
                entity for entity in self.hass.data[DATA_KEY]
                if entity.entity_id == master
            )
        )
```

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -287,10 +282,11 @@ async def async_set_volume_level(self, volume):
     async def async_join(self, master):
         """Join the group of the master player."""
+        masters = [e for e in self.hass.data[DATA_KEY] if e.entity_id == master]
         master_group = [
```
```py
master_group = next(
    (
        group
        for group in self._client.groups_available()
        if master_entity.identifier in group.clients
    )
)
```

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -287,10 +282,11 @@ async def async_set_volume_level(self, volume):
-            if master.identifier in group.clients
+            if masters[0].identifier in group.clients
         ]
         await master_group[0].add_client(self._client.identifier)
```
```suggestion
        await master_group.add_client(self._client.identifier)
```

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -58,36 +57,23 @@
+    platform.async_register_entity_service(SERVICE_SNAPSHOT, {}, "snapshot")
+    platform.async_register_entity_service(SERVICE_RESTORE, {}, "async_restore")
+    platform.async_register_entity_service(
+        SERVICE_JOIN, {vol.Required(ATTR_MASTER): cv.entity_id}, "async_join"
```
There will be an attribute error if the user sends an entity_id or a master id for an entity that is not a client but a group. The old code guarded for that. Do we need to do that still or at least raise a more specific error?

We could do that by using the alternative entity service handler api. See the bottom paragraph here:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

Eg:
```py
async def handle_async_join(entity, service_call):
	if not isinstance(entity, SnapcastClientDevice):
        raise ValueError("Incorrect entity_id. Entity is not a client.")
    await entity.async_join(service_call.data[ATTR_MASTER])
```

Then also do a similar check in the entity `async_join` method for the master entity.

BarrettLowe replied:
Still working through these...sorry for the extra request

BarrettLowe committed:
Cleanup tasks
Moved constants into separate file
Removed unnecessary logger message
Remove unnecessary schemas

BarrettLowe committed:
FIx linting errors

BarrettLowe committed:
Sort imports

BarrettLowe committed:
Update with requested change

Better - use next()

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

BarrettLowe committed:
Add guards for bad service calls

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -22,72 +21,51 @@
-CLIENT_PREFIX = "snapcast_client_"
-CLIENT_SUFFIX = "Snapcast Client"
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
We still need the platform schema. That schema is validating the config yaml, not the service data.

BarrettLowe replied:
Oh! So when I was testing without the PLATFORM_SCHEMA, everything worked fine. I assume it worked because everything was already in the config; the problem is that it just wasn't being validated? Reading [this](https://developers.home-assistant.io/docs/configuration_yaml_index) but would like a bit more clarification.

MartinHjelmare replied:
Yeah, platforms that are set up via config need a platform schema to a validate the config.

MartinHjelmare commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -22,72 +21,51 @@
+    platform.async_register_entity_service(
+        SERVICE_JOIN, {vol.Required(ATTR_MASTER): cv.entity_id}, handle_async_join
+    )
+    platform.async_register_entity_service(SERVICE_UNJOIN, {}, "async_unjoin")
```
Looks like we should guard unjoin too, and raise if the entity is not a client.

BarrettLowe replied:
Good catch

BarrettLowe committed:
Add back in platform schema

BarrettLowe committed:
Add check for unjoin service call

BarrettLowe committed:
Fix lint/format

BarrettLowe committed:
remove comma inserted by black

BarrettLowe commented:
What's the best way to run the CI tests locally? I'm using the dev container in vscode. Preferably, I could only run the formatting tools on whatever files I've changed for efficiency... Thanks

MartinHjelmare commented:
In vscode I set up the formatter to auto format on save.

To run tests I run `pytest` in the terminal from the project root. Specify the test package of the integration to just run those tests.

We also have the script `script/lazytox.py` which runs some of the CI tests locally on just the modified files.

MartinHjelmare reviewed (approved):
Looks good!
